code/game/objects/machinery/bots/cleanbot.dm

```diff
@@ -0,0 +1,169 @@
+		"Who needs an armory when we have the Roomba to secure your guns in its dustbin?",
+		"Are you perhaps unfamiliar with proper ship etiquette, my dear crayon-eating passengers?",
+	)
+	activation_animation = "cleanbot-on"
+	deactivation_animation = "cleanbot-off"
+	alter_operating_mode = TRUE
+	active_icon_state = "cleanbot1"
+
+/obj/machinery/bot/cleanbot/Initialize(mapload)
+	. = ..()
+	RegisterSignal(src, COMSIG_MOVABLE_MOVED, PROC_REF(suck_items))
+
+/obj/machinery/bot/cleanbot/process()
+	///holds the list of floors and dirs we use to evaluate which direction we're heading
+	var/list/dirtyfloors = list()
+	///used to hold what tile in the immediate area is dirtiest
+	var/highestdirtvalue = 0
+	///used to hold the dirt value of a given tile during evaluation, which is compared with highestdirtvalue
+	var/sampledirtvalue = 0
+	///used to override destination if we detect dirt or grime nearby
+	var/destdir = null
+	///controls what direction we're heading, is overridden by destdir if it has a value
+	var/newdir
+	for(var/dirn in GLOB.alldirs)
+		var/targetturf = get_step(src,dirn)
+		if(is_blocked_turf(targetturf)) //blocked turfs are not included in move calculations
+			continue
+		sampledirtvalue = calculategrime(targetturf)
+		///if the dirt value on a given tile is higher than our current dirt level, than the dirtier tile is our new destination
+		if(highestdirtvalue < sampledirtvalue)
+			highestdirtvalue = sampledirtvalue
+			destdir = dirn
+		else if(highestdirtvalue == 0)
+			dirtyfloors += dirn
+	for(var/i=1 to length(dirtyfloors))
+		newdir = pick_n_take(dirtyfloors)
+	if(!newdir && !destdir) ///if we have destdir than we have a place to go
+		say("I seem to be stuck...")
+		return
+	if(destdir)
+		step_to(src, get_step(src,destdir))
+	else
+		step_to(src, get_step(src,newdir))
+
+///return a dirt value based on how many times we detect dirty objects
+/obj/machinery/bot/cleanbot/proc/calculategrime(turf/dirtyturf)
+	var/highestdirt = 0
+	for(var/obj/dirtyobject in dirtyturf)
+		if(istype(dirtyobject, /obj/effect/decal/cleanable))
+			++highestdirt
+		if(istype(dirtyobject, /obj/item/trash))
+			++highestdirt
+	return highestdirt
+
+///Called when the roomba moves, sucks in all items held in the tile and sends them to cryo
+/obj/machinery/bot/cleanbot/proc/suck_items()
+	SIGNAL_HANDLER
+	for(var/obj/dirtyobject in loc)
+		clean(dirtyobject)
+	stuck_counter = 0
+
+///clean dirty objects and remove cleanable decals
+/obj/machinery/bot/cleanbot/proc/clean(atom/movable/O as obj|mob)
+	var/turf/currentturf = get_turf(src)
+	if(istype(O, /obj/effect/decal/cleanable))
+		playsound(loc, 'sound/effects/slosh.ogg', 25, 1)
+		currentturf.wet_floor()
+		flick("cleanbot-c", src)
+		++counter
+		if(prob(15))
+			say(pick(sentences))
+		qdel(O)
+	else if(istype(O, /obj/item/trash) || istype(O, /obj/item/shard) || istype(O, /obj/item/ammo_casing) || istype(O, /obj/effect/turf_decal/tracks/wheels/bloody))
```

Comment:
Why's this and the above two separate blocks? Plus, wetfloor/animation flick/sound should really only be once per turf and not once per cleaned thing. This should also be a list of types on the bot you can iterate through checking instead of this hardcoded set.